piccolo_api/mfa/authenticator/provider.py

```diff
@@ -0,0 +1,93 @@
+from piccolo_api.mfa.provider import MFAProvider
+
+
+class AuthenticatorProvider(MFAProvider):
+
+    def __init__(
+        self,
+        secret_table: t.Type[AuthenticatorSecret] = AuthenticatorSecret,
+        issuer_name: str = "Piccolo-MFA",
+    ):
+        """
+        :param seed_table:
+            By default, just use the out of the box ``AuthenticatorSecret``
+            table - you can specify a subclass instead if you want to override
+            certain functionality.
+        :param issuer_name:
+            This is how it will be identified in the user's authenticator app.
+
+        """
+        super().__init__(token_name="authenticator_token")
+
+        self.secret_table = secret_table
+        self.issuer_name = issuer_name
+
+    async def authenticate_user(self, user: BaseUser, code: str) -> bool:
+        """
+        The code could be a TOTP code, or a recovery code.
+        """
+        return await self.secret_table.authenticate(user_id=user.id, code=code)
+
+    async def is_user_enrolled(self, user: BaseUser) -> bool:
+        return await self.secret_table.is_user_enrolled(user_id=user.id)
+
+    async def send_code(self, *args, **kwargs):
+        """
+        Deliberately blank - the user already has the code on their phone.
+        """
+        pass
+
+    ###########################################################################
+    # Registration
+
+    async def _generate_qrcode_image(
+        self, secret: AuthenticatorSecret, email: str
+    ):
+        uri = secret.get_authentication_setup_uri(
+            email=email, issuer_name=self.issuer_name
+        )
+
+        return get_b64encoded_qr_image(data=uri)
+
+    async def get_registration_html(self, user: BaseUser) -> str:
+        """
+        When a user wants to register for MFA, this HTML is shown containing
+        instructions.
+        """
+        secret, recovery_codes = await self.secret_table.create_new(
+            user_id=user.id
+        )
+
+        qrcode_image = await self._generate_qrcode_image(
+            secret=secret, email=user.email
+        )
+
+        recovery_codes_str = "\n".join(recovery_codes)
+
+        return f"""
+            <p>Use an authenticator app like Google Authenticator to scan this QR code:</p>
+            <img src="data:image/png;base64,{qrcode_image}" />
+            <p>Copy these recovery codes and keep them safe:</p>
+            <textarea type="text">{recovery_codes_str}</textarea>
+            """  # noqa: E501
+
+    async def get_registration_json(self, user: BaseUser) -> dict:
+        """
+        When a user wants to register for MFA, the client can request a JSON
+        response, rather than HTML, if they want to render the UI themselves.
+        """
+        secret, recovery_codes = await self.secret_table.create_new(
+            user_id=user.id
+        )
+
+        qrcode_image = await self._generate_qrcode_image(
+            secret=secret, email=user.email
+        )
+
```

Comment:
I think you should add it `recovery_codes_str = "\n".join(recovery_codes)` so we get a nice recovery code format on the json response as well.

![admin_recovery_codes](https://github.com/user-attachments/assets/d0507f91-0b22-4a67-9a0c-b13a84c4d281)

I've updated the [my Piccolo Admin branch](https://github.com/piccolo-orm/piccolo_admin/compare/master...sinisaos:piccolo_admin:two_factor_auth) to use the code from this PR. If you find something interesting and the code is good enough, feel free to use it.

Reply:
> I think you should add it recovery_codes_str = "\n".join(recovery_codes) so we get a nice recovery code format on the json response as well.

Yeah, it's a tricky one - in some ways having it pre-formatted is convenient, but in others the client might want more control over how it appears (e.g. putting it in `<ul>` tag).

> I've updated the [my Piccolo Admin branch](https://github.com/piccolo-orm/piccolo_admin/compare/master...sinisaos:piccolo_admin:two_factor_auth) to use the code from this PR. If you find something interesting and the code is good enough, feel free to use it.

That's great - thanks! I'll have a look.